### PR TITLE
feat(vnav): initial check-in of VNAV, only for DECEL pseudo waypoint support

### PR DESCRIPTION
**Pull request by Benjozork**
## Summary of Changes

This merges in the WIP VNAV codebase in order to support `(DECEL)` in cFMS v1.

Most of the VNAV functionality is disabled via `VnavConfig.ts`, as follows:

```ts
export const VnavConfig = {

    /**
     * Whether to calculate climb profile
     */
    VNAV_CALCULATE_CLIMB_PROFILE: false,

    /**
     * Whether to emit ToD pseudo waypoint
     */
    VNAV_EMIT_TOD: false,

    /**
     * Whether to emit (DECEL) pseudo waypoint
     */
    VNAV_EMIT_DECEL: true,

    /**
     * VNAV descent calculation mode (NORMAL, CDA or DPO)
     */
    VNAV_DESCENT_MODE: VnavDescentMode.NORMAL,

    /**
     * Whether to emit CDA flap1/2 pseudo-waypoints (only if VNAV_DESCENT_MODE is CDA)
     */
    VNAV_EMIT_CDA_FLAP_PWP: false,

};
```

## Additional context

Discord username (if different from GitHub): someperson#4953

## Testing instructions

* Make sure `(DECEL)` pseudo-waypoint appears on ND and MCDU F-PLN page
* Make sure `(T/D)`, `(FLAP1)` and `(FLAP2)` pseudo-waypoints **do not** appear on ND and MCDU F-PLN page
* Make sure APPROACH **phase** (not mode!) auto-engages when passing `(DECEL)` when in DES **phase** (not mode!)

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by Benjozork** [comment]
Note to maintainers: I'm aware the code is a bit messy with some unused bits, but unfortunately due to the state of the system and the work on the FMS in general we can't afford either finalizing this or witholding it from cFMS v1.

File: docs/a320-simvars.md [hunk 1]
```diff
@@ -1328,6 +1328,13 @@ In the variables below, {number} should be replaced with one item in the set: {
     - Bool
     - Whether or not the GPS is used as the primary means of navigation/position determination.
 
+## Flight Management System
```
Comment (2hwk) [inline docs/a320-simvars.md:1331]: If a simvar is temporary or private then it should not appear in this .md

File: docs/a320-simvars.md [hunk 2]
```diff
@@ -1328,6 +1328,13 @@ In the variables below, {number} should be replaced with one item in the set: {
...
+
+- A32NX_FM_ENABLE_APPROACH_PHASE
+  - Bool
+  - Indicates whether the FMS should switch to APPROACH phase.
+  - **WARNING:** This is temporary and internal. Do not use.
```
Comment (2hwk) [inline docs/a320-simvars.md:1336]: If a simvar is temporary or private then it should not appear in this .md

Comment (tracernz) [reply docs/a320-simvars.md:1336]: This doc is for us developers right? The public docs on the website shouldn't have internal/private vars.

File: src/fmgc/src/guidance/vnav/DescentProfile.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            //     0,
+            //     this.isaDeviation,
+            //     this.tropoAltitude,
+            // );
+            // predDistanceTraveled += finalStepResult.distanceTraveled;
+            // predTimeElapsed += finalStepResult.timeElapsed;
+            // predFuelWeight -= finalStepResult.fuelBurned;
+
+            // this.distanceFromPposToTopOfClimb = predDistanceTraveled;
+        } while (true);
```
Comment (tracernz) [inline src/fmgc/src/guidance/vnav/DescentProfile.ts:173]: This seems dangerous to exist... I guess luckily nothing calls it yet.

File: src/shared/src/NavigationDisplay.ts [hunk 4]
```diff
@@ -44,8 +47,9 @@ export enum NdSymbolTypeFlags {
     FixInfo = 1 << 14,
     FlightPlan = 1 << 15,
     PwpDecel = 1 << 16,
-    PwpCdaFlap1White = 1 << 17,
-    PwpCdaFlap2White = 1 << 20,
+    PwpTopOfDescent = 1 << 17,
```
Comment (tracernz) [inline src/shared/src/NavigationDisplay.ts:50]: Weren't you going to change the Constraint* flags, to Constraint, and Magenta, White, Amber, so they can be used for Pwps as well?

File: src/fmgc/src/flightplanning/FlightPlanManager.ts [hunk 5]
```diff
@@ -24,13 +24,17 @@
  */
 
 import { NXDataStore } from '@shared/persistence';
-import { WaypointConstraintType } from '@fmgc/types/fstypes/FSEnums';
```
Comment (tracernz) [inline src/fmgc/src/flightplanning/FlightPlanManager.ts:27]: Can we please remove this enum from the FS typedefs? It never belonged there as it is one of ours. 😬

File: src/fmgc/src/guidance/vnav/DescentProfile.ts (hunk 3 above)
Comment (Benjozork) [reply src/fmgc/src/guidance/vnav/DescentProfile.ts]: removed

File: src/fmgc/src/efis/EfisSymbols.ts [hunk 6]
```diff
@@ -336,6 +349,17 @@ export class EfisSymbols {
                 }
             }
 
+            // Pseudo waypoints
+
+            for (const pwp of this.guidanceController.currentPseudoWaypoints.filter((it) => it)) {
+                upsertSymbol({
+                    databaseId: pwp.ident,
```
[the remark this answers is not in the document]
Comment (Benjozork) [reply src/fmgc/src/efis/EfisSymbols.ts]: fixed

File: src/shared/src/NavigationDisplay.ts (hunk 4 above)
Comment (Benjozork) [reply src/shared/src/NavigationDisplay.ts:50]: As they are not needed right now we can do it later

File: src/fmgc/src/flightplanning/FlightPlanManager.ts (hunk 5 above)
Comment (Benjozork) [reply src/fmgc/src/flightplanning/FlightPlanManager.ts:27]: done

**Comment by omarsmak** [comment]
Sorry to ask but why VNAV is disabled from the configs if it goes into experimental? 🤔

**Comment by Benjozork** [comment]
> Sorry to ask but why VNAV is disabled from the configs if it goes into experimental? 🤔

Because the disabled functionality is not finished enough to be shown yet

Nothing is stopping you from enabling them in the generated fmgc.js though.

**Comment by omarsmak** [comment]
> > Sorry to ask but why VNAV is disabled from the configs if it goes into experimental? 🤔
> 
> Because the disabled functionality is not finished enough to be shown yet
> 
> Nothing is stopping you from enabling them in the generated fmgc.js though.

Makes sense 👍

**Pull request merged**